nemo/export/trt_llm/nemo/nemo_ckpt_convert.py

```diff
@@ -0,0 +1,593 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import configparser
+import logging
+import math
+import multiprocessing
+import os
+import shutil
+import typing
+from collections import defaultdict
+from pathlib import Path
+
+import numpy as np
+import tensorstore  # this is important even though not used
+import torch
+import zarr
+from tensorrt_llm._utils import np_bfloat16, str_dtype_to_torch, torch_to_numpy
+from tqdm import tqdm
+from transformers import AutoTokenizer, GPT2Tokenizer, LlamaConfig
+
+from nemo.export.trt_llm.nemo.convert import save_weight_torch, split_and_save_weight
+from nemo.export.trt_llm.nemo.nemo import UnpackedNemoCheckpointDir, extract_layers_with_prefix, nemo_to_llm_config
+from nemo.export.trt_llm.nemo.sentencepiece_tokenizer import SentencePieceTokenizer
+
+
+LOGGER = logging.getLogger("NeMo")
+
+layer_names = {
+    "position_embedding": "embedding.position_embeddings.weight",
+    "word_embedding": "embedding.word_embeddings.weight",
+    "output_layer": "output_layer.weight",
+    "final_layernorm.weight": "final_layernorm.weight",
+    "final_layernorm.bias": "final_layernorm.bias",
+}
+
+
+def get_layer_name(layer_type: str, prefix: str):
+    layer_dict = layer_names
+    if layer_type in layer_dict:
+        return prefix + layer_dict[layer_type]
+    else:
+        raise ValueError(f"Unknown layer type {layer_type}")
+
+
+def get_layer_prefix(layer_names, is_mcore):
+    transformer_layer_prefix = None
+    model_prefix = None
+
+    for layer_name in layer_names:
+        if 'self_attention' in layer_name:
+            transformer_layer_prefix = layer_name.split('layers')[0]
+            break
+    assert transformer_layer_prefix is not None, "Cannot extract transformer layer prefix from {layer_name}"
+    if is_mcore:
+        model_prefix = transformer_layer_prefix.split('decoder')[0]
+    else:
+        model_prefix = transformer_layer_prefix.split('encoder')[0]
+    assert model_prefix is not None, "Cannot extract model prefix from {layer_name}"
+
+    return model_prefix, transformer_layer_prefix
+
+
+def get_layer_index(split_key):
+    index = 0
+    for key in split_key:
+        if key == "layers":
+            return index + 1
+        index += 1
+
+
+def rename_key(old_key: str, pp_rank: int, num_layers: int, pp_size: int):
```

Comment:
Why do need to rename keys here? Dont we already map the nemo param name to TRTLLM name inside `split_and_save_weight()`?